malleefowl/persist.py

```diff
@@ -8,38 +9,62 @@
 LOGGER = logging.getLogger("PYWPS")
 
 
-def resolve(location, place_holders, f):
+def resolve(location, f, defaults=None):
+    place_holders = re.findall('\{(.*?)\}', location)
+
+    if defaults:
+        facet_values = copy.deepcopy(defaults)
+    else:
+        facet_values = {}
+
     try:
-        nc = netCDF4.Dataset(f, 'r')
+        nc = netCDF4.Dataset(f, 'a')
     except:
-        nc = None
+        return location.format(**facet_values)
 
-    facet_values = {}
-    for facet in place_holders:
-        if hasattr(nc, facet):
-            value = getattr(nc, facet)
-        elif hasattr(nc, facet + '_id'):
-            value = getattr(nc, facet + '_id')
+    time_placeholders = ['initial_year', 'initial_month', 'initial_day',
+                         'final_year', 'final_month', 'final_day']
+    n = len(list(set(time_placeholders) - set(place_holders)))
+    # Only fetch time components if they are place_holders
+    if ('time' in nc.variables) and (n != len(time_placeholders)):
+        nctime = nc.variables['time']
+        if 'calendar' in nctime.ncattrs():
+            calendar = nctime.getncattr('calendar')
         else:
-            value = 'unknown'
-        facet_values[facet] = value
+            calendar = 'gregorian'
+        datetime_i = netCDF4.num2date(nctime[0], nctime.units, calendar)
+        facet_values['initial_year'] = str(datetime_i.year)
+        facet_values['initial_month'] = str(datetime_i.month).zfill(2)
+        facet_values['initial_day'] = str(datetime_i.day).zfill(2)
+        datetime_f = netCDF4.num2date(nctime[-1], nctime.units, calendar)
+        facet_values['final_year'] = str(datetime_f.year)
+        facet_values['final_month'] = str(datetime_f.month).zfill(2)
+        facet_values['final_day'] = str(datetime_f.day).zfill(2)
+        # Could also add support for hour, minute, second, etc.
+
+    for facet in place_holders:
+        if facet in nc.ncattrs():
+            facet_values[facet] = nc.getncattr(facet)
+        elif "{0}_id".format(facet) in nc.ncattrs():
+            facet_values[facet] = nc.getncattr("{0}_id".format(facet))
+        elif facet in defaults:
+            nc.setncattr(facet, facet_values[facet])
```

Comment:
The problem with this step inside the loop is that we are looping on the place holders but this step has nothing to do with place holders... We should be looping over all the defauts facet (not just the one referenced in the place holders).